contrib/go/src/python/pants/contrib/go/tasks/go_fetch.py

```diff
@@ -47,7 +53,27 @@ def execute(self):
     if not go_remote_libs:
       return
 
-    undeclared_deps = self._transitive_download_remote_libs(set(go_remote_libs))
+    # We cache mappings from import path to root (e.g., example.org/pkg/foo -> example.org),
+    # otherwise we may have to fetch them over the network via the meta tag protocol.
+    # Note that this caching is unversioned: The mapping is defined as "whatever meta tag is
+    # currently being served at the relevant URL", which is inherently unversioned.
+    import_root_map_path = os.path.join(self.workdir, 'pkg_root_map.txt')
+
+    if os.path.exists(import_root_map_path):
+      with open(import_root_map_path, 'r') as fp:
+        import_root_map = dict({import_path: root for import_path, root in
+                                (x.decode('utf8').strip().split('\t') for x in fp.readlines())})
+    else:
+      import_root_map = {}
+
+    undeclared_deps = self._transitive_download_remote_libs(set(go_remote_libs), import_root_map)
+
+    # Write the updated import_root_map back out.
+    safe_mkdir_for(import_root_map_path)
+    with open(import_root_map_path, 'w') as fp:
```

Comment:
Should probably use https://github.com/pantsbuild/pants/blob/master/src/python/pants/util/dirutil.py#L240 to avoid partially written/corrupted files.